File: client/src/test/scala/stasis/test/specs/unit/client/ops/recovery/stages/internal/DestagedByteStringSourceSpec.scala
package stasis.test.specs.unit.client.ops.recovery.stages.internal

import java.nio.file.Paths

import akka.actor.ActorSystem
import akka.stream.scaladsl.Source
import akka.util.ByteString
import stasis.client.analysis.Checksum
import stasis.client.api.clients.Clients
import stasis.client.ops.recovery.Providers
import stasis.client.ops.recovery.stages.internal.DestagedByteStringSource
import stasis.test.specs.unit.AsyncUnitSpec
import stasis.test.specs.unit.client.mocks._

import scala.util.control.NonFatal

class DestagedByteStringSourceSpec extends AsyncUnitSpec {
  "A DestagedByteStringSource" should "support data stream destaging" in {
    val mockStaging = new MockFileStaging()
    val mockTelemetry = MockClientTelemetryContext()

    implicit val providers: Providers = Providers(
      checksum = Checksum.MD5,
      staging = mockStaging,
      compression = MockCompression(),
      decryptor = new MockEncryption(),
      clients = Clients(
        api = MockServerApiEndpointClient(),
        core = MockServerCoreEndpointClient()
      ),
      track = new MockRecoveryTracker,
      telemetry = mockTelemetry
    )

    val original = Source.single(ByteString("original"))
    val extended = new DestagedByteStringSource(original)

    extended
      .destage(to = Paths.get("/tmp/file/one"))
      .map { _ =>
        mockStaging.statistics(MockFileStaging.Statistic.TemporaryCreated) should be(1)
        mockStaging.statistics(MockFileStaging.Statistic.TemporaryDiscarded) should be(0)
        mockStaging.statistics(MockFileStaging.Statistic.Destaged) should be(1)

        mockTelemetry.ops.recovery.entityExamined should be(0)
        mockTelemetry.ops.recovery.entityCollected should be(0)
        mockTelemetry.ops.recovery.entityChunkProcessed should be(1)
        mockTelemetry.ops.recovery.entityProcessed should be(0)
        mockTelemetry.ops.recovery.metadataApplied should be(0)
      }
  }

  it should "discard temporary staging files on failure" in {
    val mockStaging = new MockFileStaging()
    val mockTelemetry = MockClientTelemetryContext()

    implicit val providers: Providers = Providers(
      checksum = Checksum.MD5,
      staging = mockStaging,
      compression = MockCompression(),
      decryptor = new MockEncryption(),
      clients = Clients(
        api = MockServerApiEndpointClient(),
        core = MockServerCoreEndpointClient()
      ),
      track = new MockRecoveryTracker,
      telemetry = mockTelemetry
    )

    val original = Source.failed(new RuntimeException("test failure"))
    val extended = new DestagedByteStringSource(original)

    extended
      .destage(to = Paths.get("/tmp/file/one"))
      .map { _ =>
        fail("Unexpected successful result received")
      }
      .recoverWith { case NonFatal(_) =>
        mockStaging.statistics(MockFileStaging.Statistic.TemporaryCreated) should be(1)
        mockStaging.statistics(MockFileStaging.Statistic.TemporaryDiscarded) should be(1)
        mockStaging.statistics(MockFileStaging.Statistic.Destaged) should be(0)

        mockTelemetry.ops.recovery.entityExamined should be(0)
        mockTelemetry.ops.recovery.entityCollected should be(0)
        mockTelemetry.ops.recovery.entityChunkProcessed should be(0)
        mockTelemetry.ops.recovery.entityProcessed should be(0)
        mockTelemetry.ops.recovery.metadataApplied should be(0)
      }
  }

  private implicit val system: ActorSystem = ActorSystem(name = "DestagedByteStringSourceSpec")
}

File: client-android/lib/src/test/kotlin/stasis/test/client_android/lib/ops/recovery/stages/internal/DestagedByteStringSourceSpec.kt
package stasis.test.client_android.lib.ops.recovery.stages.internal

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.WordSpec
import io.kotest.matchers.shouldBe
import okio.Buffer
import stasis.client_android.lib.analysis.Checksum
import stasis.client_android.lib.api.clients.Clients
import stasis.client_android.lib.ops.recovery.Providers
import stasis.client_android.lib.ops.recovery.stages.internal.DestagedByteStringSource.destage
import stasis.test.client_android.lib.mocks.MockCompression
import stasis.test.client_android.lib.mocks.MockEncryption
import stasis.test.client_android.lib.mocks.MockFileStaging
import stasis.test.client_android.lib.mocks.MockRecoveryTracker
import stasis.test.client_android.lib.mocks.MockServerApiEndpointClient
import stasis.test.client_android.lib.mocks.MockServerCoreEndpointClient
import java.nio.file.Path
import java.nio.file.Paths

class DestagedByteStringSourceSpec : WordSpec({
    "A DestagedByteStringSource" should {
        "support data stream destaging" {
            val mockStaging = MockFileStaging()

            val providers: Providers = Providers(
                checksum = Checksum.Companion.MD5,
                staging = mockStaging,
                decompressor = MockCompression(),
                decryptor = MockEncryption(),
                clients = Clients(
                    api = MockServerApiEndpointClient(),
                    core = MockServerCoreEndpointClient()
                ),
                track = MockRecoveryTracker()
            )

            val original = Buffer().write("original".toByteArray())

            original.destage(
                to = Paths.get("/tmp/file/one"),
                providers = providers
            )

            mockStaging.statistics[MockFileStaging.Statistic.TemporaryCreated] shouldBe (1)
            mockStaging.statistics[MockFileStaging.Statistic.TemporaryDiscarded] shouldBe (0)
            mockStaging.statistics[MockFileStaging.Statistic.Destaged] shouldBe (1)
        }

        "discard temporary staging files on failure" {
            val mockStaging = object : MockFileStaging() {
                override suspend fun destage(from: Path, to: Path) {
                    throw RuntimeException("Test failure")
                }
            }

            val providers: Providers = Providers(
                checksum = Checksum.Companion.MD5,
                staging = mockStaging,
                decompressor = MockCompression(),
                decryptor = MockEncryption(),
                clients = Clients(
                    api = MockServerApiEndpointClient(),
                    core = MockServerCoreEndpointClient()
                ),
                track = MockRecoveryTracker()
            )

            val original = Buffer().write("original".toByteArray())

            val e = shouldThrow<RuntimeException> {
                original.destage(
                    to = Paths.get("/tmp/file/one"),
                    providers = providers
                )
            }

            e.message shouldBe ("Test failure")

            mockStaging.statistics[MockFileStaging.Statistic.TemporaryCreated] shouldBe (1)
            mockStaging.statistics[MockFileStaging.Statistic.TemporaryDiscarded] shouldBe (1)
            mockStaging.statistics[MockFileStaging.Statistic.Destaged] shouldBe (0)
        }
    }
})
